### PR TITLE
[Issue 11412] Add shutdownGracefully for ManagedLedgerFactoryImpl

### LINKED ISSUE #11412: When open ledger encountered factory shutdown, callback may never be called back.
**Describe the bug**
For the moment that ManagedLedgerFactoryImpl.open() runs into MetaStoreImpl.getManagedLedgerInfo() and waiting for the future returned by store.get() to complete, if ManagedLedgerFactoryImpl.shutdown() is invoked then, the executor will be shutdown, and thus the future will be unable to run following stage after it's completion, as the executor will reject. Unfortunately, the exceptionally() also requires the executor, and the failure callback won't be called, open() will never return.

**To Reproduce**
Refer to the following test:

```java
public class ManagedLedgerTest {

    @Test
    public void openEncounteredShutdown() throws Exception {
        final String ledgerName = UUID.randomUUID().toString();
        final long version = 0;
        final long createTimeMillis = System.currentTimeMillis();

        MetadataStore metadataStore = mock(MetadataStore.class);
        given(metadataStore.get(any())).willAnswer(inv -> {
            String path = inv.getArgumentAt(0, String.class);
            if (path == null) {
                throw new IllegalArgumentException("Path is null.");
            }
            if (path.endsWith(ledgerName)) { // ledger
                MLDataFormats.ManagedLedgerInfo.Builder mli = MLDataFormats.ManagedLedgerInfo.newBuilder()
                        .addLedgerInfo(0, MLDataFormats.ManagedLedgerInfo.LedgerInfo.newBuilder()
                                .setLedgerId(0)
                                .setEntries(0)
                                .setTimestamp(System.currentTimeMillis()));
                Stat stat = new Stat(path, version, createTimeMillis, createTimeMillis);
                return CompletableFuture.supplyAsync(() -> {
                    try {
                        Thread.sleep(10000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    System.out.println("store.get return with ml");
                    return Optional.of(new GetResult(mli.build().toByteArray(), stat));
                });

            } else if (path.contains(ledgerName)) { // cursor
                MLDataFormats.ManagedCursorInfo.Builder mci = MLDataFormats.ManagedCursorInfo.newBuilder()
                        .setCursorsLedgerId(-1)
                        .setMarkDeleteLedgerId(0)
                        .setMarkDeleteLedgerId(-1);
                Stat stat = new Stat(path, version, createTimeMillis, createTimeMillis);
                return CompletableFuture.supplyAsync(() -> {
                    try {
                        Thread.sleep(10000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    System.out.println("store.get return with mc");
                    return Optional.of(new GetResult(mci.build().toByteArray(), stat));
                });

            } else {
                throw new IllegalArgumentException("Invalid path: " + path);
            }
        });
        given(metadataStore.put(anyString(), any(), any())).willAnswer(inv -> {
            Optional<Long> expectedVersion = inv.getArgumentAt(2, Optional.class);
            return CompletableFuture.supplyAsync(() -> {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                return new Stat(inv.getArgumentAt(0, String.class), expectedVersion.orElse(0L) + 1, createTimeMillis, System.currentTimeMillis());
            });
        });
        given(metadataStore.getChildren(anyString())).willAnswer(inv -> {
            return CompletableFuture.supplyAsync(() -> {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                return Collections.singletonList("cursor");
            });
        });

        BookKeeper bookKeeper = mock(BookKeeper.class);
        LedgerHandle ledgerHandle = mock(LedgerHandle.class);
        LedgerHandle newLedgerHandle = mock(LedgerHandle.class);
        OrderedExecutor executor = OrderedExecutor.newBuilder().name("Test").build();
        given(bookKeeper.getMainWorkerPool()).willReturn(executor);
        doAnswer(inv -> {
            Thread.sleep(100);
            AsyncCallback.OpenCallback cb = inv.getArgumentAt(3, AsyncCallback.OpenCallback.class);
            cb.openComplete(0, ledgerHandle, inv.getArgumentAt(4, Object.class));
            return null;
        }).when(bookKeeper).asyncOpenLedger(anyLong(), any(), any(), any(), any());
        doAnswer(inv -> {
            Thread.sleep(100);
            AsyncCallback.CreateCallback cb = inv.getArgumentAt(5, AsyncCallback.CreateCallback.class);
            cb.createComplete(0, newLedgerHandle, inv.getArgumentAt(6, Object.class));
            return null;
        }).when(bookKeeper).asyncCreateLedger(anyInt(), anyInt(), anyInt(), any(), any(), any()/*callback*/, any(), any());

        ManagedLedgerFactoryImpl factory = new ManagedLedgerFactoryImpl(metadataStore, bookKeeper);
        CountDownLatch countDownLatch = new CountDownLatch(1);
        CompletableFuture.runAsync(() -> {
            try {
                ManagedLedger ml = factory.open(ledgerName);
            } catch (Throwable e) {
                e.printStackTrace();
            } finally {
                countDownLatch.countDown();
            }
        });

        Thread.sleep(5000);
        System.out.println("Shutdown...");
        factory.shutdown();

        if (!countDownLatch.await(60, TimeUnit.SECONDS)) {
            fail("open() not returned in time.");
        }
    }
}
```

**Expected behavior**
ManagedLedgerFactoryImpl.open()/asyncOpen() should fail with error instead of hang forever.

**Additional context**
VERSION: v2.8.0

### PR DESCRIPTION
**Pull request by Jason918**
Fixes #11412 

### Motivation

see #11412 

### Modifications

1.  Add Method shutdownGracefully for the user case, which will tries its best to clean up resources and make sure all callbacks is called.
2.  Add closed indicating that shutdown is called. If shutdown is called twice, `ManagedLedgerFactoryClosedException` will be thrown.
3. Fix some test cases which make duplicate shutdown calls to ManagedLedgerFactory by using both `@Cleanup("shutdown") ` and  `factory.shutdown()`.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change to `shutdown` is already covered by existing tests, such as org.apache.pulsar.transaction.coordinator.test.MockedBookKeeperTestCase#tearDown.

and 

This change added tests and can be verified as follows:
  - `org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryShutdownTest#openEncounteredShutdown`

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

#### For contributor

For this PR, do we need to update docs?
- No, fix some conner case, it does not affect anything to do with normal users.

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 1]
```diff
@@ -472,8 +484,87 @@ void close(ManagedLedger ledger) {
...
+
+        statsTask.cancel(true);
+        flushCursorsTask.cancel(true);
+
+        int closedLedgerNum = 0;
+
+        while (!ledgers.isEmpty()) { //use while loop for ledgers added during shutdown.
+            List<String> ledgerNames = new ArrayList<>(this.ledgers.keySet());
+            int numLedgers = ledgerNames.size();
+            final CountDownLatch latch = new CountDownLatch(numLedgers);
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:502]: How about make the `shutdownGracefully()` asynchronously?

Comment (Jason918) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:502]: Great idea, shutdownAsync is more flexible.

**Comment by Jason918** [comment]
@codelipenghui  I have changed to shutdownAsync, PTAL.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 2]
```diff
@@ -472,8 +485,107 @@ void close(ManagedLedger ledger) {
...
+    public CompletableFuture<Void> shutdownAsync() throws ManagedLedgerException {
+        if (closed) {
+            throw new ManagedLedgerException.ManagedLedgerFactoryClosedException();
+        }
+        closed = true;
+
+        statsTask.cancel(true);
+        flushCursorsTask.cancel(true);
+
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:497]:
```suggestion
        List<CompletableFuture<Void>> futures = new ArrayList<>(numLedgers);
```

**Review by codelipenghui — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Pull request merged**